src/LeaseManager.h

```diff
@@ -0,0 +1,157 @@
+     * with cleaning during a cleaning pass.
+     */
+    class LeaseCleaner : public WorkerTimer {
+      public:
+        explicit LeaseCleaner(Context* context,
+                              LeaseManager* leaseManager);
+        virtual void handleTimerEvent();
+
+        LeaseManager* leaseManager;
+      private:
+        DISALLOW_COPY_AND_ASSIGN(LeaseCleaner);
+    };
+
+    /// Monitor-style lock
+    SpinLock mutex;
+    typedef std::lock_guard<SpinLock> Lock;
+
+    /// Shared information about the server.
+    Context* context;
+
+    /// Used to provided a guaranteed monotonically increasing number (even
+    /// across crashes) that (mostly) advances in sync with real time.  This
+    /// Cluster Time is used to define lease expiration times.
+    CoordinatorClusterClock clock;
+
+    /// Represents the largest leaseId that was issued from this module.  The
+    /// module guarantees that no leaseId will be issued more than once.  The
+    /// next leaseId issued should be ++lastIssuedLeaseId.
+    uint64_t lastIssuedLeaseId;
+
+    /// This is the largest leaseId that has been pre-allocated in external
+    /// storage.  Pre-allocating leaseIds allows this module to respond to
+    /// requests for new leases without waiting for an external storage
+    /// operation.  This module must never issue a leaseId greater than this
+    /// value.  To guarantee this value is recovered after a crash we must make
+    /// sure that maxAllocatedLeaseId is never removed (i.e. has its lease
+    /// freed).  In the normal case, continual pre-allocations will make sure
+    /// that this value runs ahead of lastIssuedLeaseId and thus cannot have
+    /// its corresponding lease freed.
+    uint64_t maxAllocatedLeaseId;
+
+    /// Maps from leaseId to its leaseTerm (the time after which the lease may
+    /// expire).  This is used to quickly service requests about a lease's
+    /// liveness.  This structure is updated whenever a lease is added, renewed,
+    /// or removed.
+    typedef std::unordered_map<uint64_t, uint64_t> LeaseMap;
+    LeaseMap leaseMap;
+
+    /// Structure to define the entries in the ExpirationOrderSet.
+    struct ExpirationOrderElem {
+        uint64_t leaseTerm;     // ClusterTime after which the lease can expire.
+        uint64_t leaseId;       // Id of the lease.
+
+        /**
+         * The operator < is overridden to implement the
+         * correct comparison for the expirationOrder.
+         */
+       bool operator<(const ExpirationOrderElem& elem) const {
+           return leaseTerm < elem.leaseTerm ||
+               (leaseTerm == elem.leaseTerm && leaseId < elem.leaseId);
+       }
+    };
+
+    /// This ordered set keeps track of the order in which leases can expire.
+    /// This structure allows the module to quickly determine what leases can be
+    /// expired next.  This structure should always be updated to refect changes
+    /// to the LeaseMap structure when the LeaseMap is updated.
```

Comment:
I don't understand how this keeps track of order: aren't sets unordered? Let's discuss.

Also, I think we may have discussed this previously, but why not just ensure that expiration times are unique, so you don't have to deal with the possibility of several leases expiring at the same time?